smartplaylist: allow exporting item fields, rename m3u8 output to extm3u

mgoltzsche opened the pull request:
## Description

Allow generating extm3u playlists so that they contain additional item fields such as the `id`.
    
The feature is required by the [webm3u plugin](https://github.com/mgoltzsche/beets-webm3u) (M3U server) to transform playlists using a request based item URI template which may require additional fields such as the `id`, e.g. `beets:library:track;$id`.

## To Do

<!--
- If you believe one of below checkpoints is not required for the change you
  are submitting, cross it out and check the box nonetheless to let us know.
  For example: - [x] ~Changelog~
- Regarding the changelog, often it makes sense to add your entry only once
  reviewing is finished. That way you might prevent conflicts from other PR's in
  that file, as well as keep the chance high your description fits with the
  latest revision of your feature/fix.
- Regarding documentation, bugfixes often don't require additions to the docs.
- Please remove the descriptive sentences in braces from the enumeration below,
  which helps to unclutter your PR description.
-->

- [x] Documentation. (If you've added a new command-line flag, for example, find the appropriate page under `docs/` to describe it.)
- [x] Changelog. (Add an entry to `docs/changelog.rst` to the bottom of one of the lists near the top of the document.)
- [x] Tests. (Very much encouraged but not strictly required.)

Serene-Arc commented:
Thank you for this PR! The code seems fine but might I ask that you add some more to the docs? Perhaps the keys to the attributes most likely to be used and an explanation for when they might be required, such as the web service you specified.

JOJ0 commented:
Hi @mgoltzsche, great addition the smartplaylist and very interesting plugin you wrote. Please also add a link to your plugin's repo to this section of the docs when you are coming back to working on this PR: https://beets.readthedocs.io/en/latest/plugins/index.html#interoperability

mgoltzsche commented:
I rebased the PR now to resolve the merge conflict, added an entry to the interoperability section of the plugins/index.rst, added more info to the smartplaylist.rst.
Also, I changed the option name from `attributes` to `fields`.
Unfortunately, the gh cache of the PR seems to be inconsistent since it doesn't show my force-pushed commit :disappointed:....

mgoltzsche commented:
Okay, as described [here](https://stackoverflow.com/questions/45626986/updating-pull-request-after-review-changes-commit-doesnt-show-up-in-the-pull-re) re-selecting `beetbox:master` as base branch within the GitHub GUI updates the PR and triggers CI but now the diff also includes all other commits on the master branch since this PR was created :disappointed:. However, my changes are still within a single commit which you could re-review.

JOJ0 commented:
Hi, I don't find time to really look into the code you wrote but I think @Serene-Arc approved that already. Just one nitpick about terminology. I'm not sure if m3u8 is the right one, I think that just tells that characters could be unicode and not ascii. At least according to wikipedia. Maybe that is not accurate anyway? I don't know, what was your source?

I think it should rather be "Extended M3U", and maybe the output option should call that `extm3u` as well. It looks like this was incorrect before your changes here! Not sure if we should fix that here now if we don't want to break existing users configs though! :-/

https://en.wikipedia.org/wiki/M3U#Extended_M3U

https://en.wikipedia.org/wiki/M3U#M3U8

And I'm still not sure if I interpret that wikipedia article entirely correct...

mgoltzsche commented:
Indeed, I added that m3u8 format feature in a previous PR and indeed, it should be called extm3u instead and I was going to address that within a separate PR.
Eventually I'll add a commit to this PR to change that tomorrow...

mgoltzsche commented:
Since I rebased this PR, added documentation as @Serene-Arc requested and changed the option name from `attributes` to `fields`, technically she has to re-review it :sweat_smile:.

JOJ0 commented:
> Indeed, I added that m3u8 format feature in a previous PR and indeed, it should be called extm3u instead and I was going to address that within a separate PR.
> 
> Eventually I'll add a commit to this PR to change that tomorrow...

Ah ol thanks for letting me know, so m3u8 format option us maybe only existing since a couple of months? In that case yeah, the impact might not be too massive on the userbase.

I agree that changing is a good idea.

mgoltzsche commented:
> so m3u8 format option is maybe only existing since a couple of months?

Yes, the `--output=m3u|m3u8` option was introduced in #5053.

mgoltzsche commented:
Okay, now I renamed the `output` option value `m3u8` to `extm3u`.
Also, I added a link to the [Beetstream plugin](https://github.com/BinaryBrain/Beetstream) to the interoperability section of the plugin documentation, given that its playlist endpoints [now also serve the smartplaylist generated playlists](https://github.com/BinaryBrain/Beetstream/pull/13).

mgoltzsche commented:
I updated my previous comment after I made a few more changes to the documentation.
Please re-review!

mgoltzsche commented:
@JOJ0 are you okay with having the webm3u plugin link within the "Other plugins" section of the documentation as opposed the interoperability section? Otherwise, do you think the webrouter plugin should also be listed in the interoperability documentation section?

JOJ0 commented:
Ah yes please put everything into the Other Plugins section. Seems like we file all of externally hosted plugins there. My suggestion here was wrong. I'm sorry: https://github.com/beetbox/beets/pull/5121#issuecomment-1982637778

mgoltzsche commented:
Rebased to resolve merge conflict within changelog.

JOJ0 commented:
Sorry for the massive delay in reviewing, I wanted to do this earlier. Code is fine, there is only one thing that I suggest to improve: Please add one or two example lines to the docs. How will a playlist entry look like with a specific config.

JOJ0 commented on test/plugins/test_smartplaylist.py:
```diff
@@ -241,6 +241,58 @@ def test_playlist_update_output_m3u8(self):
+        self.assertEqual(
+            content,
+            b"#EXTM3U\n"
+            + b'#EXTINF:300 id="456" genre="Fake Genre",Fake Artist - fake Title\n'
```
Something like this: Add a config example and the resulting line on the playlist file to the docs.

JOJ0 reviewed (changes requested):
Sorry for the massive delay in reviewing, I wanted to do this earlier. Code is fine, there is only one thing that I suggest to improve: Please add one or two example lines to the docs. How will a playlist entry look like with a specific config?

mgoltzsche commented:
Okay, I added an example configuration and extm3u playlist to the documentation now. Please re-review!

mgoltzsche commented:
After some problems with the docs linter now the PR is ready for review again.
Also, I made the documentation mention the autogenre plugin.

JOJ0 commented:
BTW I like the tiny PlaylistItem class, something like that could also help with my never finished project of providing a general m3u handling solution in beets. Currently it's rather unspectacular, it has save/load functionality only: 

oops oof-topic alert :-)
